src/oops/SOLID/openClosePrinciple/before/client/CalculateTaxesClient.java

```diff
@@ -5,31 +5,34 @@
 import java.util.Locale;
 
 import oops.SOLID.openClosePrinciple.before.employees.Employee;
+import oops.SOLID.openClosePrinciple.before.employees.FullTimeEmployee;
+import oops.SOLID.openClosePrinciple.before.employees.Intern;
+import oops.SOLID.openClosePrinciple.before.employees.PartTimeEmployee;
 import oops.SOLID.openClosePrinciple.before.persistence.EmployeeRepository;
+import oops.SOLID.openClosePrinciple.before.taxes.FullTimeEmployeeTaxCalculator;
+import oops.SOLID.openClosePrinciple.before.taxes.InternTaxCalculator;
+import oops.SOLID.openClosePrinciple.before.taxes.PartTimeEmployeeTaxCalculator;
 import oops.SOLID.openClosePrinciple.before.taxes.TaxCalculator;
+import oops.SOLID.openClosePrinciple.before.taxes.TaxCalculatorFactory;
 
 
 public class CalculateTaxesClient {
     public static void main(String[] args) {
-       
-        EmployeeRepository repository = new EmployeeRepository();
-
-        // Grab employees
-        List<Employee> employees = repository.findAll();
+        final TaxCalculatorFactory taxCalculatorFactory = getTaxCalculatorFactory();
+        final EmployeeRepository repository = new EmployeeRepository();
 
         // Calculate taxes
-        Locale locale = new Locale("en", "US");
-        NumberFormat currencyFormatter = NumberFormat.getCurrencyInstance(locale);
-        TaxCalculator taxCalculator = new TaxCalculator();
-
-        double totalTaxes = 0;
-        for (Employee employee: employees){
+        double totalTaxes = repository.findAll().stream()
+            .map(employee -> taxCalculatorFactory.getTaxCalculator(employee.getClass().getTypeName())
+                .calculate(employee))
+            .reduce(0.0, Double::sum);
+    }
 
-            // compute individual tax
-            double tax = taxCalculator.calculate(employee);
-            String formattedTax = currencyFormatter.format(tax);
-            // add to company total taxes
-            totalTaxes += taxCalculator.calculate(employee);
-        }
+    private static TaxCalculatorFactory getTaxCalculatorFactory() {
+        final TaxCalculatorFactory factory = new TaxCalculatorFactory();
+        factory.register(FullTimeEmployee.class.getTypeName(), new FullTimeEmployeeTaxCalculator());
+        factory.register(PartTimeEmployee.class.getTypeName(), new PartTimeEmployeeTaxCalculator());
+        factory.register(Intern.class.getTypeName(), new InternTaxCalculator());
+        return factory;
```

Comment:
Shouldn't this register logic be present in TaxCalculatorFactory class? I think this is currently violating SRP!

Reply:
Oh! I will explain in detail.

There are two ways to do this
[Approach 1] is what you suggest, letting factory know what all instances of `TaxCalculator`s are available on compile time (Ether holding static map or writing a big switch case construct inside factory.) and having hard dependancy between implementations and factory.

[Approach 2] is registering all instances of `TaxCalculator` instances on deploy/bootstrap time, and letting factory only serve these instances on request.

The pros/cons of these approaches will not be apparent when number of implementations is very small (which is true in this case). However once your application starts growing and number of implementations increase, it will be hard to manage these instances statically inside factory.

That is where the [Approach 2] shines. Basically "Single Responsibility" of `factory` here is only to serve instances and not to know what instances it is serving. By segregating responsibility this way, I am offloading headache of configuring factory to some other class. In this case I added this logic to `Client` class since this is where all logic of bootstrapping resides, however I could have created a `TaxCalculatorFactoryConfiguration` class, and put this logic there. 

This way the implementations can be registered agnostically on boot time (through dependency injection) and factory doesn't need to change when you add or remove implementations.

Hope this helps. Otherwise you can ping me separately, I have some other examples in my projects.

Reply:
Thanks for explaining the approach in detail. Really appreciate that!

> factory doesn't need to change when you add or remove implementations
true, but now client needs to change when you add or remove implementations.

> however I could have created a TaxCalculatorFactoryConfiguration class, and put this logic there.
yeah, I think this would have been better.

Anyways, I would love to see more such implementations to get a hang of this.

Reply:
Right! In Spring like environment it would make sense to configure this factory in configuration class (like you mentioned) and expose it as bean. That way you can ensure that factory is singleton